### ISSUE TITLE
Add automatic heroic  discovery 

### ISSUE THREAD

**Issue by LeSnake04** [mtkennerly/ludusavi#94]
The program should automatically detect the root folder for heroic games launcher.

This would mean the programs needs to understand the heroic configuration.

**Comment by mtkennerly** [comment]
Does the Heroic launcher have its own save directory as well? I see that it has some cloud save sync functionality, but if the original local saves are still at the GOG/Epic paths, then you would just configure the GOG/Epic roots in Ludusavi.

**Comment by LeSnake04** [comment]
I tried adding it as gog / epic but it dint work. Seems like they have their own file structure.

I can send a file tree or a zip without games (on my pc i just have game installed outside the heroic folder - this data will be usefull for detecting externaly installed games)

**Comment by LeSnake04** [comment]
also note heroic will often be used as flatpak

**Comment by LeSnake04** [comment]
This will have to be combined with the file `/home/lesnake/.var/app/com.heroicgameslauncher.hgl/config/heroic/GamesConfig/APPNAME_OF_GAME.json`

Example: `/home/lesnake/.var/app/com.heroicgameslauncher.hgl/config/heroic/GamesConfig/1942942943.json`
```json
{
  "1942942943": {
    "altLegendaryBin": "",
    "altGogdlBin": "",
    "addDesktopShortcuts": false,
    "addStartMenuShortcuts": false,
    "autoInstallDxvk": false,
    "autoInstallVkd3d": false,
    "checkForUpdatesOnStartup": false,
    "customWinePaths": [],
    "defaultInstallPath": "/data/epic",
    "defaultWinePrefix": "/home/lesnake/Games/Heroic/Prefixes",
    "egsLinkedPath": "",
    "language": "en",
    "maxRecentGames": 5,
    "maxWorkers": 0,
    "minimizeOnLaunch": false,
    "nvidiaPrime": false,
    "otherOptions": "",
    "showFps": false,
    "showUnrealMarket": false,
    "useGameMode": false,
    "wineCrossoverBottle": "Heroic",
    "winePrefix": "/home/lesnake/Games/Heroic/Prefixes/Bridge-Constructor-Portal",
    "wineVersion": {
      "bin": "",
      "name": "Default Wine - Not Found",
      "type": "wine"
    }
  },
  "version": "v0",
  "explicit": true
}
```

**Comment by mtkennerly** [comment]
Thanks for the detailed info! That sounds like it would work.

**Comment by LeSnake04** [comment]
No problem. I will be happy to test it whenever its ready.

Note:
these root folders should be added by default in the release with the heroic type: 
- `~/.var/app/com.heroicgameslauncher.hgl/config/heroic/`
- `~/.config/heroic/`

**Comment by LeSnake04** [comment]
Also I am currently learning rust. Once I am done with my current project, I can try to implement it.

**Comment by mtkennerly** [comment]
That would be great! Contributions are always welcome :) For reference, it would need to be added here: https://github.com/mtkennerly/ludusavi/blob/c308170517c04e079e621d53d464a74025fec33b/src/config.rs#L261

**Comment by LeSnake04** [comment]
It would help when, whenever you have some  tme, could install heroic on windows and tell where the configs are located. I can maybe start on this in 1-2 weeks depending how smooth my other project goes.

**Comment by mtkennerly** [comment]
My equivalent of your `/home/lesnake/.var/app/com.heroicgameslauncher.hgl/config/heroic` is `C:\Users\mtken\AppData\Roaming\heroic`. I'm using the portable version (Heroic_Portable_2.4.0.exe), if that matters. However:

* After installing Cave Story+ from Epic, `GamesConfig` contained only `1dea8a6ddb544842a58e4b5c8675ff58.log` and no JSON file.
  * When I opened the in-app settings for the game, then it generated a `1dea8a6ddb544842a58e4b5c8675ff58.json` containing an empty object: `{}`
  * After launching the game from Heroic, the JSON file stayed as `{}`.
* `gog_store` is empty (which makes sense, but I don't see any equivalent folder for Epic)

I tried another game (The Alto Collection) with a custom install location, and it ended up with an empty object as well, so I'm not sure how Heroic tracks install folders. Maybe it queries the information from Epic/Legendary directly and only has to store the info in the JSON files for GOG?

I tried Tyrian 2000 from GOG, and it produced a `GamesConfig/1207658901.json` containing `{}`, as well as a `gog_store/installed.json` containing the install path:

```json
{
	"installed": [
		{
			"platform": "windows",
			"executable": "",
			"install_path": "C:\\Users\\mtken\\Games\\Heroic\\Tyrian 2000",
			"install_size": "22.74 MiB",
			"is_dlc": false,
			"version": "3.0",
			"appName": "1207658901",
			"installedWithDLCs": false,
			"language": "en-US",
			"versionEtag": "\"3640869511\"",
			"buildId": "52095636268561745"
		}
	]
}
```

In summary:

* The process you described would work for Linux + GOG.
* For Epic, the install folders are not obvious.
* For Windows, we can ignore `GamesConfig`.

We might need to open a ticket in the Heroic repository to ask what a reliable way to do this would be. They may need to expose some API that Ludusavi could use to get the required info generically for Epic, GOG, and any other stores they might support in the future.

**Comment by LeSnake04** [comment]
interestingly, gog_store tracks epic and gog game for me.

**Comment by LeSnake04** [comment]
Update: I found  the relevant folder for epic on linux: 
$HOME/.var/app/com.heroicgameslauncher.hgl/config/legendary/installed.json
```
{
  "Herring": {
    "app_name": "Herring",
    "base_urls": [
      "https://download.epicgames.com/Builds/Org/o-ws89bsbfb8trgdl3zmerugg4rsuvvt/50c7fef4b8a44187a4c71ca5686b41b3/default",
      "https://download2.epicgames.com/Builds/Org/o-ws89bsbfb8trgdl3zmerugg4rsuvvt/50c7fef4b8a44187a4c71ca5686b41b3/default",
      "https://download3.epicgames.com/Builds/Org/o-ws89bsbfb8trgdl3zmerugg4rsuvvt/50c7fef4b8a44187a4c71ca5686b41b3/default",
      "https://download4.epicgames.com/Builds/Org/o-ws89bsbfb8trgdl3zmerugg4rsuvvt/50c7fef4b8a44187a4c71ca5686b41b3/default",
      "https://epicgames-download1.akamaized.net/Builds/Org/o-ws89bsbfb8trgdl3zmerugg4rsuvvt/50c7fef4b8a44187a4c71ca5686b41b3/default",
      "https://fastly-download.epicgames.com/Builds/Org/o-ws89bsbfb8trgdl3zmerugg4rsuvvt/50c7fef4b8a44187a4c71ca5686b41b3/default"
    ],
    "can_run_offline": true,
    "egl_guid": "",
    "executable": "Hue.exe",
    "install_path": "/home/david/Games/Heroic/Hue",
    "install_size": 477828681,
    "install_tags": [],
    "is_dlc": false,
    "launch_parameters": "",
    "manifest_path": null,
    "needs_verification": false,
    "platform": "Windows",
    "prereq_info": null,
    "requires_ot": false,
    "save_path": null,
    "title": "Hue",
    "version": "1.0"
  }
}
```

**Comment by LeSnake04** [comment]
I decided that intead of building a heroic parser into ludusavi, i am gonna make a lib for that for us to impletemnt here. my plan is that we just supply the paths to heroic/legendary and gives back a struct containg info including the path to the wineprefix.

**Comment by mtkennerly** [comment]
That sounds like a good idea! Thanks for working on this one \:D

**Comment by LeSnake04** [comment]
also I found out that you only have empty entries in GamesConfig, because the files in there store the wine configuration. Since vereything is native on windows, there are none. I also have a few empty ones for gog games with native linux versions.

**Comment by sluedecke** [comment]
Hi all, do you have a recent status on this issue, maybe some unfinished code one could work on?  I'd like to contribute since this is an important feature IMO.  GOG Linux native games are already well supported, but games installed with wine / proton in individual prefixes require adding the individual roots.

**Comment by sluedecke** [comment]
I created a very rough implementation here: https://github.com/sluedecke/ludusavi/tree/feature/94-heroic-game-roots

It detects the heroic configuration folder on Linux, checks `gog_store/installed.json` and looks for prefixes in the GameConfig/*.json files.  These are added to the list of roots on the fly.

**Comment by sluedecke** [comment]
This is my background research (for the sake of completeness): 
[feature-94.md](https://github.com/mtkennerly/ludusavi/files/9737198/feature-94.md)

**Comment by mtkennerly** [comment]
That looks like great progress :D Thanks for investigating.

Regarding this part from the "algorithm" section:

```
match platform {
      windows:
        [...]
        add found prefix as possible install_dir
```

I'm not sure that the prefix needs to be checked as an install dir. Maybe we could use the `wine_prefix` argument to `prelude::scan_game_for_backup`, or (at a lower level) we could just check it in `prelude::parse_paths` like we do for Steam + Proton currently. I'm not sure which would work best.

**Comment by sluedecke** [comment]
You are right and the implementation I did went another way.  I removed the algorithm section to avoid confusion.  I also updated the comments in config.rs to reflect the code.  The changes I've made can be summarized as follows:

* Added new store type `Store::HeroicConfig` 
* `Config::find_missing_roots` has some Linux defaults for the heroic config folder
* `Config::expanded_roots` checks for such store and adds the wine/prefixes of all games it finds to the returned value

After that the normal program flow happens and it seems to work in the cases I described in the https://github.com/mtkennerly/ludusavi/files/9738607/feature-94.md

**Comment by sluedecke** [comment]
@LeSnake04 can you test my implementation and check if it works for you?

**Comment by sluedecke** [comment]
Moved the code into `config::detect_heroic_gog_roots` which populates `config::heroic_gog_roots`.  cli and app use this field to provide a `wine_prefix` to `scan_game_for_backup`.

This works, but the type change of `wine_prefix` from `&Option<StrictPath>` to `&Option<&StrictPath>` is not well handled right now (marked with TODOs)

**Comment by LeSnake04** [comment]
![Screenshot_20221009_223601](https://user-images.githubusercontent.com/47243761/194778341-e4a45b19-6f52-4944-9e3c-b10e079d9075.png)

I dont see a heroicconfig to add the games. find roots also doesnt work.

**Comment by LeSnake04** [comment]
It works!
![Screenshot_20221009_223601](https://user-images.githubusercontent.com/47243761/194779432-8520f30c-ce98-43fa-8b9c-eaeca0e7e72d.png)

**Comment by LeSnake04** [comment]
It doesnt seem to detect epic games.

**Comment by sluedecke** [comment]
@LeSnake04 thank you for giving it a spin!  Epic games are not yet handled at all since I don't have any of those, not even an epic account.  Can you help me out with some - credentials removed of course - example configuration files?  Or alternatively recommend some small, free games I can use for testing on my machine?

**Comment by sluedecke** [comment]
OK, some data points:

* legendary config stored in `~/.config/legendary` / `~/.var/app/com.heroicgameslauncher.hgl/config/legendary`
* game .app_name and .title can be found in `~/.config/legendary/installed.json` but also in `~/.config/legendary/metadata/<app_name>.json`.  I will go for the first
* heroic config for specific game stored by heroic in `CONFIGDIR/GamesConfig/<app_name>.json`, format the same as for gog games

So I'll add a `config::detect_heroic_legendary_roots` and try it out.

**Comment by LeSnake04** [comment]
Details se here https://github.com/mtkennerly/ludusavi/issues/94#issuecomment-1217874144

**Comment by mtkennerly** [comment]
If you can't find a game with an exact match, you can fall back to `prelude::normalize_title` to do a rough match.

Pseudocode:

```rust
let normalized_to_official: HashMap<_> = manifest.keys().map(|title| (normalize_title(title), title)).collect();

for candidate in heroic_games {
    let normalized = normalize_title(candidate.title);
    if let Some(official) = normalized_to_official.get(normalized) {
        // we found a match
    }
}
```

**Comment by sluedecke** [comment]
@mtkennerly thanks!  I am not sure where to best include this code.   Do you have some advice?

Using it during game detection in config seems to be a good place, but in order to work, config needs to know the manifest.  Only the cli/app know the manifest so far.

Refactoring the code for heroic into a dedicated ~~class~~struct seems to be a much cleaner way.  It can then be used by cli/app like:

```rust
// heroic.rs
struct HeroicGameList {
    heroic_roots: std::collections::HashMap<String, StrictPath>,
}

impl HeroicGamesList {
  pub fn new(roots: &Vec<RootsConfig, manifest: &Manifest>) -> Self {
    // initialize and run game detection here
  }
  pub fn get_heroic_root(&self, title: &String) -> Option<StrictPath> {
    // return value from self.heroic_roots
  }
}
...

// app.rs / cli.rs
// during app::start_backup / cli::Subcommand::Backup
...
let s = 
s.detect_heroic_games();
...
scan_game_for_backup(..., s.get_heroic_root(&key), ...)
...
```

What do you think?

**Comment by mtkennerly** [comment]
I like that approach. It's very similar to how `InstallDirRanking` works. I would suggest using its design as the basis for a `PrefixRanking` struct:

```rust
#[derive(Clone, Default)]
pub struct PrefixRanking(std::collections::HashMap<(RootsConfig, String), StrictPath>);

impl PrefixRanking {
    pub fn get(&self, root: &RootsConfig, name: &str) -> Option<&StrictPath> {}

    pub fn scan(roots: &[RootsConfig], manifest: &crate::manifest::Manifest, subjects: &[String]) -> Self {}
}
```

The `scan` function can just inspect the Heroic roots and ignore any other roots for now. By keying on root + game, it can handle multiple Heroic roots.

It probably doesn't need to keep track of scores like `InstallDirRanking` does, but we can always add that if necessary.

**Comment by LeSnake04** [comment]
It seems to work. EGS games got detected on my system with no apparant issues. Thanks for the hard work.

**Comment by sluedecke** [comment]
@mtkennerly ludusavi now makes use of `normalize_title` as you suggest in `heroic::memorize_game_root` and it works fine for 'Slain: Back from Hell'.  I also did some refactoring and println! changes to make the output more readable.

There are still some todos left, but I am pretty sure that the main work has been done for this now.

**Comment by sluedecke** [comment]
Regarding Steam Deck: though it works for me, it does not work well for a installation of heroic done via flatpak.

#### A) Installation with standard Linux tools

I installed heroic manually as a normal package with pacman on the deck.  This involves disabling read only for the Steam Deck file system and is something the majority of users will not do.  Still it creates a situation on the deck which is similar to a normal installation on any other machine and heroic configuration is found in `~/.config/heroic`.
In this situation ludusavi finds the save games in all the known / expected locations.

#### B) Flatpak installation

However, if one chooses to install heroic via flatpak, heroic configuration ends up in `~/.var/app/com.heroicgameslauncher.hgl/config/heroic`.  The problem is with some Linux native games, e.g. Terraria and Spiritfarer.  They then store their save games in `.var/app/com.heroicgameslauncher.hgl/data/Terraria`  (instead of `~/.local/share/Terraria`) and `.var/app/com.heroicgameslauncher.hgl/config/unity3d/Thunder Lotus Games/Spiritfarer`. And ludusavi does not find these.  Windows games seem to work fine so far.

So for a flatpak installation more work needs to be done.

@mtkennerly if you agree, I would work to finish this item for situation A and create a new issue to extend it to properly handle B once A is done.

**Comment by LeSnake04** [comment]
The thing is: Most games use /home/user even when installed trough heroic / steam Flatpak. Only some actually use the flatpak sandbox of the store.

IMO the way to go for B is a Root called "flatpak sandbox" thats under the box is handled like a Home, but with .config being replaced with config and .local/share replaced by data.

**Comment by sluedecke** [comment]
Thanks for the feedback!  I will open the pull request for A and create a new issue for B then.

The situation with which game uses which folder to store things really is a bit "complicated" since there is no real [standard](https://xkcd.com/927/).

**Comment by sluedecke** [comment]
Please note that I removed flatpak locations from the pull request, I will reintroduce them in #142
